File: TheScopes/Content/Python/TheScopesEngine.py
import math
import datetime

class TheScopesEngine:
    """The Scopes calculation engine - provides astrology, numerology, biorhythm, and moon phase calculations."""
    
    ZODIAC_SIGNS = [
        (1, 20, 2, 18, "Aquarius"),
        (2, 19, 3, 20, "Pisces"),
        (3, 21, 4, 19, "Aries"),
        (4, 20, 5, 20, "Taurus"),
        (5, 21, 6, 20, "Gemini"),
        (6, 21, 7, 22, "Cancer"),
        (7, 23, 8, 22, "Leo"),
        (8, 23, 9, 22, "Virgo"),
        (9, 23, 10, 22, "Libra"),
        (10, 23, 11, 21, "Scorpio"),
        (11, 22, 12, 21, "Sagittarius"),
        (12, 22, 1, 19, "Capricorn"),
    ]
    
    LIFE_PATH_INTERPRETATIONS = {
        1: "The Leader - Independent, ambitious, innovative",
        2: "The Peacemaker - Cooperative, diplomatic, sensitive",
        3: "The Communicator - Creative, expressive, optimistic",
        4: "The Builder - Practical, disciplined, hardworking",
        5: "The Adventurer - Freedom-loving, versatile, curious",
        6: "The Nurturer - Responsible, caring, harmonious",
        7: "The Seeker - Analytical, spiritual, introspective",
        8: "The Achiever - Ambitious, powerful, material-focused",
        9: "The Humanitarian - Compassionate, idealistic, generous",
        11: "The Master Intuitive - Spiritual insight, inspiration",
        22: "The Master Builder - Visionary, practical idealism",
        33: "The Master Teacher - Compassion, spiritual guidance",
    }
    
    EXPRESSION_INTERPRETATIONS = {
        1: "Natural leader, independent, creative vision",
        2: "Diplomatic, cooperative, harmony-seeking",
        3: "Creative expression, communication, joy",
        4: "Practical builder, organized, reliable",
        5: "Versatile, adventurous, freedom of expression",
        6: "Nurturing, responsible, beauty-loving",
        7: "Analytical, spiritual depth, introspective",
        8: "Ambitious, material success, authority",
        9: "Humanitarian, compassionate, global vision",
    }
    
    SOUL_URGE_INTERPRETATIONS = {
        1: "Desires independence and personal achievement",
        2: "Desires partnership and emotional harmony",
        3: "Desires creative expression and joy",
        4: "Desires stability, order, and security",
        5: "Desires freedom, adventure, and variety",
        6: "Desires love, family, and responsibility",
        7: "Desires spiritual understanding and solitude",
        8: "Desires material success and recognition",
        9: "Desires to serve humanity and make a difference",
    }
    
    @staticmethod
    def reduce_to_single_digit(number):
        """Reduce a number to single digit (keeping master numbers 11, 22, 33)."""
        while number > 9 and number not in (11, 22, 33):
            number = sum(int(d) for d in str(number))
        return number
    
    @staticmethod
    def get_letter_value(letter):
        """Get numerology value of a letter (A=1, B=2, ... I=9, J=1, etc.)."""
        letter = letter.upper()
        if not letter.isalpha():
            return 0
        return (ord(letter) - ord('A')) % 9 + 1
    
    @staticmethod
    def is_vowel(letter):
        """Check if a letter is a vowel."""
        return letter.upper() in 'AEIOU'
    
    @classmethod
    def calculate_sun_sign(cls, birth_month, birth_day):
        """Calculate zodiac sun sign from birth month and day."""
        for start_m, start_d, end_m, end_d, name in cls.ZODIAC_SIGNS:
            if start_m == end_m:
                if birth_month == start_m and start_d <= birth_day <= end_d:
                    return name
            else:
                if (birth_month == start_m and birth_day >= start_d) or \
                   (birth_month == end_m and birth_day <= end_d):
                    return name
        return "Unknown"
    
    @classmethod
    def calculate_moon_phase(cls, year, month, day):
        """Calculate current moon phase name and illumination percentage."""
        # Known new moon: January 6, 2000
        ref_date = datetime.date(2000, 1, 6)
        target_date = datetime.date(year, month, day)
        days_since_ref = (target_date - ref_date).days
        
        lunar_cycle = 29.53059
        cycle_position = days_since_ref % lunar_cycle
        if cycle_position < 0:
            cycle_position += lunar_cycle
        
        phase_percent = cycle_position / lunar_cycle
        phase = phase_percent * 8.0
        
        if phase < 0.5 or phase >= 7.5:
            phase_name = "New Moon"
        elif phase < 1.5:
            phase_name = "Waxing Crescent"
        elif phase < 2.5:
            phase_name = "First Quarter"
        elif phase < 3.5:
            phase_name = "Waxing Gibbous"
        elif phase < 4.5:
            phase_name = "Full Moon"
        elif phase < 5.5:
            phase_name = "Waning Gibbous"
        elif phase < 6.5:
            phase_name = "Last Quarter"
        else:
            phase_name = "Waning Crescent"
        
        return phase_name, phase_percent
    
    @classmethod
    def calculate_life_path(cls, birth_date_str):
        """Calculate life path number from birth date string (YYYY-MM-DD)."""
        parts = birth_date_str.split('-')
        if len(parts) != 3:
            return 0, "Invalid date format. Use YYYY-MM-DD."
        
        year = int(parts[0])
        month = int(parts[1])
        day = int(parts[2])
        
        year_sum = cls.reduce_to_single_digit(year)
        month_sum = cls.reduce_to_single_digit(month)
        day_sum = cls.reduce_to_single_digit(day)
        
        life_path = cls.reduce_to_single_digit(year_sum + month_sum + day_sum)
        interpretation = cls.LIFE_PATH_INTERPRETATIONS.get(life_path, "Unique path")
        
        return life_path, interpretation
    
    @classmethod
    def calculate_expression_number(cls, full_name):
        """Calculate expression number from full name."""
        total = sum(cls.get_letter_value(c) for c in full_name if c.isalpha())
        expression = cls.reduce_to_single_digit(total)
        interpretation = cls.EXPRESSION_INTERPRETATIONS.get(expression, "Unique expression")
        return expression, interpretation
    
    @classmethod
    def calculate_soul_urge(cls, full_name):
        """Calculate soul urge number from full name (vowels only)."""
        total = sum(cls.get_letter_value(c) for c in full_name if cls.is_vowel(c))
        soul_urge = cls.reduce_to_single_digit(total)
        interpretation = cls.SOUL_URGE_INTERPRETATIONS.get(soul_urge, "Unique soul desire")
        return soul_urge, interpretation
    
    @classmethod
    def calculate_biorhythm(cls, days_alive):
        """Calculate biorhythm percentages for physical, emotional, and intellectual cycles."""
        physical = math.sin(2 * math.pi * days_alive / 23) * 100
        emotional = math.sin(2 * math.pi * days_alive / 28) * 100
        intellectual = math.sin(2 * math.pi * days_alive / 33) * 100
        return physical, emotional, intellectual
    
    @classmethod
    def calculate_all(cls, full_name, birth_date_str):
        """Calculate all scope readings at once."""
        parts = birth_date_str.split('-')
        year, month, day = int(parts[0]), int(parts[1]), int(parts[2])
        
        birth_date = datetime.date(year, month, day)
        today = datetime.date.today()
        days_alive = (today - birth_date).days
        
        sun_sign = cls.calculate_sun_sign(month, day)
        moon_phase_name, moon_phase_percent = cls.calculate_moon_phase(today.year, today.month, today.day)
        life_path, life_path_interp = cls.calculate_life_path(birth_date_str)
        expression, expression_interp = cls.calculate_expression_number(full_name)
        soul_urge, soul_urge_interp = cls.calculate_soul_urge(full_name)
        physical, emotional, intellectual = cls.calculate_biorhythm(days_alive)
        
        return {
            "sun_sign": sun_sign,
            "moon_phase": moon_phase_name,
            "moon_phase_percent": moon_phase_percent,
            "life_path": life_path,
            "life_path_interpretation": life_path_interp,
            "expression": expression,
            "expression_interpretation": expression_interp,
            "soul_urge": soul_urge,
            "soul_urge_interpretation": soul_urge_interp,
            "physical": round(physical, 1),
            "emotional": round(emotional, 1),
            "intellectual": round(intellectual, 1),
            "days_alive": days_alive,
        }


# Register as a UE Python module
# Usage: from TheScopesEngine import TheScopesEngine
# result = TheScopesEngine.calculate_all("Patricia Susur", "1990-05-15")

File: TheScopes/Source/TheScopes/Private/TheScopesEngine.cpp
#include "TheScopesEngine.h"
#include "Internationalization/Regex.h"

FString UTheScopesEngine::CalculateSunSign(int32 BirthMonth, int32 BirthDay)
{
    // Zodiac sign date ranges (month, start_day, end_day, sign_name)
    struct FZodiacSign
    {
        int32 StartMonth;
        int32 StartDay;
        int32 EndMonth;
        int32 EndDay;
        FString Name;
    };

    const FZodiacSign Signs[] = {
        {1, 20, 2, 18, TEXT("Aquarius")},
        {2, 19, 3, 20, TEXT("Pisces")},
        {3, 21, 4, 19, TEXT("Aries")},
        {4, 20, 5, 20, TEXT("Taurus")},
        {5, 21, 6, 20, TEXT("Gemini")},
        {6, 21, 7, 22, TEXT("Cancer")},
        {7, 23, 8, 22, TEXT("Leo")},
        {8, 23, 9, 22, TEXT("Virgo")},
        {9, 23, 10, 22, TEXT("Libra")},
        {10, 23, 11, 21, TEXT("Scorpio")},
        {11, 22, 12, 21, TEXT("Sagittarius")},
        {12, 22, 1, 19, TEXT("Capricorn")}
    };

    for (const FZodiacSign& Sign : Signs)
    {
        if (Sign.StartMonth == Sign.EndMonth)
        {
            // Same month range (e.g., Aquarius: Jan 20 - Feb 18)
            if (BirthMonth == Sign.StartMonth && BirthDay >= Sign.StartDay && BirthDay <= Sign.EndDay)
            {
                return Sign.Name;
            }
        }
        else
        {
            // Cross-month range
            if ((BirthMonth == Sign.StartMonth && BirthDay >= Sign.StartDay) ||
                (BirthMonth == Sign.EndMonth && BirthDay <= Sign.EndDay))
            {
                return Sign.Name;
            }
        }
    }

    return TEXT("Unknown");
}

void UTheScopesEngine::CalculateMoonPhase(int32 Year, int32 Month, int32 Day, FString& PhaseName, float& PhasePercent)
{
    // Simple moon phase calculation based on known new moon reference
    // Known new moon: January 6, 2000 at 18:14 UTC
    // Lunar cycle: 29.53059 days
    
    // Calculate days since reference new moon
    int32 DaysSinceRef = 0;
    
    // Simple day count from year 2000
    DaysSinceRef = (Year - 2000) * 365 + (Year - 2000) / 4; // Approximate
    for (int32 m = 1; m < Month; m++)
    {
        if (m == 1 || m == 3 || m == 5 || m == 7 || m == 8 || m == 10 || m == 12)
            DaysSinceRef += 31;
        else if (m == 4 || m == 6 || m == 9 || m == 11)
            DaysSinceRef += 30;
        else if (m == 2)
        {
            if (Year % 4 == 0 && (Year % 100 != 0 || Year % 400 == 0))
                DaysSinceRef += 29;
            else
                DaysSinceRef += 28;
        }
    }
    DaysSinceRef += Day - 6; // Reference day is Jan 6
    
    // Adjust for leap years between 2000 and target year
    int32 LeapDays = (Year - 2000) / 4 - (Year - 2000) / 100 + (Year - 2000) / 400;
    DaysSinceRef += LeapDays;
    
    // Calculate position in lunar cycle
    const float LunarCycle = 29.53059f;
    float CyclePosition = FMath::Fmod((float)DaysSinceRef, LunarCycle);
    if (CyclePosition < 0) CyclePosition += LunarCycle;
    
    PhasePercent = CyclePosition / LunarCycle;
    
    // Determine phase name
    float Phase = PhasePercent * 8.0f;
    if (Phase < 0.5f || Phase >= 7.5f)
        PhaseName = TEXT("New Moon");
    else if (Phase < 1.5f)
        PhaseName = TEXT("Waxing Crescent");
    else if (Phase < 2.5f)
        PhaseName = TEXT("First Quarter");
    else if (Phase < 3.5f)
        PhaseName = TEXT("Waxing Gibbous");
    else if (Phase < 4.5f)
        PhaseName = TEXT("Full Moon");
    else if (Phase < 5.5f)
        PhaseName = TEXT("Waning Gibbous");
    else if (Phase < 6.5f)
        PhaseName = TEXT("Last Quarter");
    else
        PhaseName = TEXT("Waning Crescent");
}

int32 UTheScopesEngine::ReduceToSingleDigit(int32 Number)
{
    while (Number > 9 && Number != 11 && Number != 22 && Number != 33)
    {
        int32 Sum = 0;
        while (Number > 0)
        {
            Sum += Number % 10;
            Number /= 10;
        }
        Number = Sum;
    }
    return Number;
}

int32 UTheScopesEngine::GetLetterValue(TCHAR Letter)
{
    Letter = FChar::ToUpper(Letter);
    if (Letter < 'A' || Letter > 'Z') return 0;
    
    int32 Value = (Letter - 'A') % 9 + 1;
    return Value;
}

bool UTheScopesEngine::IsVowel(TCHAR Letter)
{
    Letter = FChar::ToUpper(Letter);
    return Letter == 'A' || Letter == 'E' || Letter == 'I' || Letter == 'O' || Letter == 'U';
}

void UTheScopesEngine::CalculateLifePath(const FString& BirthDate, int32& LifePathNumber, FString& Interpretation)
{
    // Parse birth date (YYYY-MM-DD format)
    TArray<FString> Parts;
    BirthDate.ParseIntoArray(Parts, TEXT("-"), true);
    
    if (Parts.Num() != 3)
    {
        LifePathNumber = 0;
        Interpretation = TEXT("Invalid date format. Use YYYY-MM-DD.");
        return;
    }
    
    int32 Year = FCString::Atoi(*Parts[0]);
    int32 Month = FCString::Atoi(*Parts[1]);
    int32 Day = FCString::Atoi(*Parts[2]);
    
    // Reduce each component
    int32 YearSum = ReduceToSingleDigit(Year);
    int32 MonthSum = ReduceToSingleDigit(Month);
    int32 DaySum = ReduceToSingleDigit(Day);
    
    // Add together and reduce
    LifePathNumber = ReduceToSingleDigit(YearSum + MonthSum + DaySum);
    
    // Interpretations
    switch (LifePathNumber)
    {
        case 1: Interpretation = TEXT("The Leader - Independent, ambitious, innovative"); break;
        case 2: Interpretation = TEXT("The Peacemaker - Cooperative, diplomatic, sensitive"); break;
        case 3: Interpretation = TEXT("The Communicator - Creative, expressive, optimistic"); break;
        case 4: Interpretation = TEXT("The Builder - Practical, disciplined, hardworking"); break;
        case 5: Interpretation = TEXT("The Adventurer - Freedom-loving, versatile, curious"); break;
        case 6: Interpretation = TEXT("The Nurturer - Responsible, caring, harmonious"); break;
        case 7: Interpretation = TEXT("The Seeker - Analytical, spiritual, introspective"); break;
        case 8: Interpretation = TEXT("The Achiever - Ambitious, powerful, material-focused"); break;
        case 9: Interpretation = TEXT("The Humanitarian - Compassionate, idealistic, generous"); break;
        case 11: Interpretation = TEXT("The Master Intuitive - Spiritual insight, inspiration"); break;
        case 22: Interpretation = TEXT("The Master Builder - Visionary, practical idealism"); break;
        case 33: Interpretation = TEXT("The Master Teacher - Compassion, spiritual guidance"); break;
        default: Interpretation = TEXT("Unique path"); break;
    }
}

void UTheScopesEngine::CalculateExpressionNumber(const FString& FullName, int32& ExpressionNumber, FString& Interpretation)
{
    int32 Total = 0;
    
    for (TCHAR Char : FullName)
    {
        if (FChar::IsAlpha(Char))
        {
            Total += GetLetterValue(Char);
        }
    }
    
    ExpressionNumber = ReduceToSingleDigit(Total);
    
    switch (ExpressionNumber)
    {
        case 1: Interpretation = TEXT("Natural leader, independent, creative vision"); break;
        case 2: Interpretation = TEXT("Diplomatic, cooperative, harmony-seeking"); break;
        case 3: Interpretation = TEXT("Creative expression, communication, joy"); break;
        case 4: Interpretation = TEXT("Practical builder, organized, reliable"); break;
        case 5: Interpretation = TEXT("Versatile, adventurous, freedom of expression"); break;
        case 6: Interpretation = TEXT("Nurturing, responsible, beauty-loving"); break;
        case 7: Interpretation = TEXT("Analytical, spiritual depth, introspective"); break;
        case 8: Interpretation = TEXT("Ambitious, material success, authority"); break;
        case 9: Interpretation = TEXT("Humanitarian, compassionate, global vision"); break;
        default: Interpretation = TEXT("Unique expression"); break;
    }
}

void UTheScopesEngine::CalculateSoulUrge(const FString& FullName, int32& SoulUrgeNumber, FString& Interpretation)
{
    int32 Total = 0;
    
    for (TCHAR Char : FullName)
    {
        if (IsVowel(Char))
        {
            Total += GetLetterValue(Char);
        }
    }
    
    SoulUrgeNumber = ReduceToSingleDigit(Total);
    
    switch (SoulUrgeNumber)
    {
        case 1: Interpretation = TEXT("Desires independence and personal achievement"); break;
        case 2: Interpretation = TEXT("Desires partnership and emotional harmony"); break;
        case 3: Interpretation = TEXT("Desires creative expression and joy"); break;
        case 4: Interpretation = TEXT("Desires stability, order, and security"); break;
        case 5: Interpretation = TEXT("Desires freedom, adventure, and variety"); break;
        case 6: Interpretation = TEXT("Desires love, family, and responsibility"); break;
        case 7: Interpretation = TEXT("Desires spiritual understanding and solitude"); break;
        case 8: Interpretation = TEXT("Desires material success and recognition"); break;
        case 9: Interpretation = TEXT("Desires to serve humanity and make a difference"); break;
        default: Interpretation = TEXT("Unique soul desire"); break;
    }
}

void UTheScopesEngine::CalculateBiorhythm(int32 DaysAlive, float& Physical, float& Emotional, float& Intellectual)
{
    const float PI = 3.14159265358979f;
    
    // Biorhythm cycles:
    // Physical: 23 days
    // Emotional: 28 days
    // Intellectual: 33 days
    
    // Calculate using sine wave: sin(2 * PI * daysAlive / cycleLength)
    Physical = FMath::Sin(2.0f * PI * (float)DaysAlive / 23.0f) * 100.0f;
    Emotional = FMath::Sin(2.0f * PI * (float)DaysAlive / 28.0f) * 100.0f;
    Intellectual = FMath::Sin(2.0f * PI * (float)DaysAlive / 33.0f) * 100.0f;
}
